### PR TITLE
feat: v1

### PR DESCRIPTION
**Pull request by ryanycw**
## Summary

Implement the main entities and handlers with the current contract.

- Challenge: Everything related to challenge
- User: Everything related to user
- UserChallenge: Mapping between challenge and user
- CheckInRecord: Everything related to checkin
- UserChallengeCheckInRecord: Mapping between checkin data and user + challenge

## Enhancement

There are some enhancement that can be achieved in the next version

### This can be achieved by tracking the token transfer event
- totalDonation: BigInt #TODO uint256

### These can be achieved by tracking the handleUserOp event
- createdChallenges: [Bytes!]! #TODO uint256[] 
- totalCreatedChallenges: BigInt #TODO uint256
- settledChallenges: [Bytes!]! #TODO uint256[]
- totalSettledChallenges: BigInt #TODO uint256

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -94,3 +94,59 @@ type Settle @entity(immutable: true) {
...
+  totalCheckIns: BigInt! # uint256
+  totalClaims: BigInt! # uint256
+  totalSucceedUsers: BigInt! # uint256
+  totalFailedUsers: BigInt! # uint256
+  totalDonation: BigInt #TODO uint256
+}
+
+type User @entity {
+  id: Bytes! # address
+  joinedChallenges: [UserChallenge!] @derivedFrom(field: "user") # Challenge[]
```
Comment (antoncoding) [inline schema.graphql]: we can combined `joinedChallenges`, `claimedChallenges`, `settledChallenges` into as single `history: [UserChallenge!]`

or is it something you plan on doing?

File: schema.graphql [hunk 2]
```diff
@@ -94,3 +94,59 @@ type Settle @entity(immutable: true) {
...
+  settledChallenges: [BigInt!] #TODO uint256[]
+  totalSettledChallenges: BigInt #TODO uint256
+  totalSucceedChallenges: BigInt! # uint256
+}
+
+type UserChallenge @entity {
+  id: Bytes! # uint256
+  user: User! # address
+  challengeId: Challenge! # uint256
+  status: Int8! # enum
```
Comment (antoncoding) [inline schema.graphql:142]: I think we can add a `checkIns: [CheckInDigest]` here! So in the future we can easily display checkin graphs ...etc

File: schema.graphql [hunk 3]
```diff
@@ -94,3 +94,59 @@ type Settle @entity(immutable: true) {
...
+}
+
+type CheckInDigest @entity {
```
Comment (antoncoding) [inline schema.graphql]: maybe call it CheckIn. the term `digest` is for ERC712 for hashing

File: schema.graphql (hunk 1 above)
Comment (ryanycw) [reply schema.graphql]: I think it's the naming issue, I've renamed it to make it clearer. I want to make joinedChallenges -> history. And claimedChallenges and settledChallenges is more like the personal record of the claiming history and the challenges that the user help settled. Does that make sense to you?

File: schema.graphql [hunk 4]
```diff
@@ -94,3 +94,64 @@ type Settle @entity(immutable: true) {
...
+  id: Bytes! # user + challengeId
+  user: User! # address
+  challengeId: Challenge! # uint256
+  checkIns: [UserChallengeCheckInRecord!] @derivedFrom(field: "userChallenge")
+  status: Int8! # enum
+}
+
+type CheckInRecord @entity {
+  id: Bytes! # checkInData
+  usedInChallenge: [UserChallengeCheckInRecord!] @derivedFrom(field: "checkInRecord")
```
Comment (antoncoding) [inline schema.graphql:147]: 🧠 🧠 🧠

**Review by antoncoding — APPROVED** [review]
LGTM! 👍 👍 👍 
FEATURE COMPLETE ACHIEVED IN V1!

**Pull request merged**